doc/compat/latch.adoc

```diff
@@ -0,0 +1,133 @@
+////
+Copyright 2023 Christian Mazakas
+Distributed under the Boost Software License, Version 1.0.
+https://www.boost.org/LICENSE_1_0.txt
+////
+
+[#latch]
+# Latch
+:idprefix: latch_
+
+The header `<boost/core/latch.hpp>` implements, in a portable way, the C++20 
+`<latch>` header.
+
+`latch` is a single-use barrier which counts downwards, enabling synchronization
+between a set of threads. The counter can be manually decremented by any value,
+but decrementing below zero produces UB. The maximum number of waiters is
+specified to be `boost::core::latch::max()`.
+
+## Example
+
+```cpp
+std::ptrdiff_t const num_threads = 16;
+boost::core::latch l(num_threads);
+
+std::vector<std::thread> threads;
+for (int i = 0; i < num_threads; ++i) {
+  threads.emplace_back([&l] {
+    // do some preliminary work here...
+
+    // block until all threads have reached this statement
+    l.arrive_and_wait();
+
+    // continue with further work...
+  });
+}
+
+for (auto& t: threads) { t.join(); }
+```
+
+## Synopsis
+
+```cpp
+namespace boost
+{
+namespace compat
+{
+
+class latch {
+  explicit latch(std::ptrdiff_t expected);
+
+  latch(latch const &) = delete;
+  latch &operator=(latch const &) = delete;
+
+  ~latch() = default;
+
+  void count_down(std::ptrdiff_t n = 1);
+  bool try_wait() const noexcept;
+  void wait() const;
+  void arrive_and_wait(std::ptrdiff_t n = 1);
+
+  static constexpr std::ptrdiff_t max() noexcept;
+};
+
+} // namespace compat
+} // namespace boost
+```
+
+## Constructors
+
+### Counter Constructor
+
+```cpp
+explicit latch(std::ptrdiff_t expected);
+```
+
+[horizontal]
+Preconditions:;; `expected >= 0 && expected \<= boost::core::latch::max()`.
+Effects:;; Constructs a latch with an internal counter value of `expected`.
+
+### Copy Constructor
+
+```cpp
+latch(latch const &) = delete;
+```
+
+`latch` is not copyable or movable.
+
+## Member Functions
+
+### count_down
+
+```cpp
+void count_down(std::ptrdiff_t n = 1);
+```
+
+[horizontal]
+Preconditions:;; `n` must not be larger than the current internal count.
+Effects:;; Decrements the internal counter by `n`.
+
+### try_wait
+
+```cpp
+bool try_wait() const noexcept;
+```
+
+Returns a boolean indicating whether or not the latch's internal count has reached 0 (`true`) or not (`false`).
+
+### wait
+
+```cpp
+void wait() const;
+```
+
+Blocks the current thread until the internal counter has reached 0.
+
+
+### arrive_and_wait
+
+```cpp
+void arrive_and_wait(std::ptrdiff_t n = 1);
+```
+
+[horizontal]
+Preconditions:;; `n` must not be larger than the current internal count.
+Effects:;; Decrements the internal counter by `n` and if the counter non-zero, blocks the current thread.
+
+### max
+
+```cpp
+static constexpr std::ptrdiff_t max() noexcept;
+```
+
+Returns an implementation-defined number representing the maximum amount of waiters. Currently INT_MAX.
```

Comment:
Same question as before; why not PTRDIFF_MAX?

Reply:
Originally, I only checked what libstdc++ did which was use INT_MAX and I wanted to keep parity with that. But both MS STL and libc++ seem to use PTRDIFF_MAX.

I don't really have a good reason for sticking with INT_MAX other than "libstdc++ does it". We can change this to PTRDIFF_MAX to keep us in line with the other stdlib implementations if you'd like.

I suppose in my mind, I was thinking a user may not need a latch with a count that high but who knows. I imagine this class to be used where the count == thread count and it'd be pretty tough to have that many threads.

Reply:
Since the type is `ptrdiff_t`, having the max value be something else is odd and would need better justification than "but the user doesn't need that many threads."